src/core/game.test.js

```diff
@@ -100,23 +101,20 @@ test('serpentine setup phases', () => {
       phases: [
         {
           name: 'first setup round',
-          turnOrder: {
-            first: () => '0',
-            next: (G, ctx) => (+ctx.currentPlayer + 1) % ctx.numPlayers + '',
-          },
+          turnOrder: TurnOrder.ONETIME,
         },
         {
           name: 'second setup round',
-          turnOrder: {
-            first: (G, ctx) => ctx.numPlayers - 1 + '',
-            next: (G, ctx) => (+ctx.currentPlayer - 1) % ctx.numPlayers + '',
+          turnOrder: TurnOrder.ONETIME,
+          onPhaseBegin: (G, ctx) => {
+            ctx.playerOrder = [3, 2, 1, 0];
```

Comment:
All the move and flow functions are meant to just return a new version of `G`. `ctx` is a read-only object.

We should find another way to update the play-order here. One way might be to create a new event (similar to `endTurn`) whose only purpose is to update `playOrder`.

Reply:
But that's not ideal either because you can't call it from within a move.

Overall, I still think my suggestion above for a `TurnOrder.CUSTOM` is simpler and doesn't add any additional complexity to `flow.js` which is already getting a bit hard to follow. Is there any type of game that you want to implement that you can't using that?

Reply:
A lot of games are using a simple round-robin turn order. Sometimes the only requirement is to rotate the first user from phase to phase, but always with the same default order (i.e 0,1,2,3), which you already covered with the startingPlayerToken example in game.test.js. But this is always assuming that the player order is the joining order. 

Terra Mystica, Gaia Project and others have a different turn order mechanism. Usually the setup phase is requiring something like 0,1,2,3,3,2,1,0 , or if some races are in play 0,1,2,2,1,0,0,3.
Then during the play phases, the pass order will determine the new play order.
Some boardgames are basing the new turn order on the victory points or the resources on the previous phase.

When you are playing tournaments, usually you have an initial draft phase, where you bet victory points to select your turn order.

The only option I'm seeing, if we want to keep the current implementation and reject this pull request, is to store the playOrder in G and use it in a TurnOrder.CUSTOM. What do you prefer?

Reply:
Right.

I guess my question is does `TurnOrder.CUSTOM` cover all these cases? If it does, I think I would prefer that because it doesn't touch anything in `flow.js`.

Reply:
I guess it does. I will remove playOrder from ctx and implement it as a G property

Reply:
What if we use G.playerOrder and keep using the TurnOrder.DEFAULT ,  TurnOrder.REVERSE, TurnOrder.SKIP for the turn behaviour ?  Something like 
 ````
  DEFAULT: {
    first: (G, ctx) => {
      if (G.playerOrder !== undefined && G.playerOrder.length > 0) {
        return G.playerOrder[0] + '';
      } else {
        return ctx.currentPlayer;
      }
    },
````

Reply:
I think it's better to be more explicit with a new turn order rather than have the existing turn orders be affected by a field in G.

Reply:
Yes, I agree with you.My only concern is that we will end up adding `TurnOrder.CUSTOM_DEFAULT`, `TurnOrder.CUSTOM_SKIP` and the `TurnOrder.CUSTOM_ONETIME`

Reply:
Oh, I see.

In that case, how about we parameterize `DEFAULT`, `SKIP` and `ONETIME`?

In the case that no additional argument is provided, they behave as they do today. In case an argument is passed, a field in `G` by the same name is used to determine the turn order instead.

Reply:
Actually, after thinking about this a bit more, how about we use something like the proposal in #142 in order to set `playerOrder` in `ctx` (through an event called `Event.setPlayerOrder()`), and then use it like you are currently?

That way, we don't leverage anything in `G` while still providing the developer with a way to manipulate the turn order (which is stored entirely in `ctx`) from within a move.